hashjoin.go

```diff
@@ -0,0 +1,393 @@
+		hashTable: hashTable,
+	}
+}
+
+func (builder *hashJoinGroupBuilder) insertBatch(flow dataFlow, eqColIdx int, outCols []int, batchSize int) {
+	builder.hashTable.growGroup(batchSize)
+
+	eqCol := flow.b[eqColIdx].(intColumn)[:batchRowLen]
+
+	for i := 0; i < batchSize; i++ {
+		builder.hashTable.keys[i + builder.totalSize + 1] = eqCol[i]
+	}
+
+	for idx, colIdx := range outCols {
+		outCol := flow.b[colIdx].(intColumn)[:batchRowLen]
+		builder.spread(outCol, idx, batchSize)
+	}
+
+	builder.totalSize += batchSize
+}
+
+func (builder *hashJoinGroupBuilder) spread(outCol intColumn, valIdx int, batchSize int) {
+	valCol := builder.hashTable.values[valIdx]
+
+	for i:= 0; i < batchSize; i++ {
+		valCol[i + builder.totalSize + 1] = outCol[i]
+	}
+}
+
+func (builder *hashJoinGroupBuilder) computeBucket() {
+	builder.bucket = make([]int, builder.totalSize)
+
+	for i := 0; i < builder.totalSize; i++ {
+		builder.bucket[i] = builder.hashTable.hashInt(builder.hashTable.keys[i + 1])
+	}
+}
+
+func (builder *hashJoinGroupBuilder) insertKeys() {
+	builder.hashTable.growNext(builder.totalSize)
+
+	for i := 0; i < builder.totalSize; i++ {
+		builder.hashTable.insertKey(builder.bucket[i], i + 1)
+	}
+}
+
+func (builder *hashJoinGroupBuilder) start(source ExecOp, eqColIdx int, outCols []int) {
+	for{
+		flow := source.Next()
+		batchSize := flow.n
+
+		if batchSize == 0 {
+			break
+		}
+
+		builder.insertBatch(flow, eqColIdx, outCols, batchSize)
+	}
+
+	builder.computeBucket()
+	builder.insertKeys()
+}
+
+// hashJoinerIntInt performs a hashJoin joining on two integer columns where the
+// left table represents the build relation. It does not work with N - N joins.
+type hashJoinerIntInt struct {
+	hashTable *hashTableInt
+
+	left ExecOp
+	right ExecOp
+
+	leftEqColIdx  int
+	rightEqColIdx int
+
+	leftCols []int
+	rightCols []int
+
+	outFlow dataFlow
+}
+
+func (hashJoiner *hashJoinerIntInt) Init() {
+	nOutputCols := len(hashJoiner.leftCols) + len(hashJoiner.rightCols)
+	if nOutputCols == 0 {
+		panic("no output cols")
+	}
+
+	hashJoiner.outFlow.b = make(batch, nOutputCols)
+	for i := range hashJoiner.outFlow.b {
+		hashJoiner.outFlow.b[i] = make(intColumn, batchRowLen)
+	}
+
+	// todo(changangela): we currently assume build relation has unique join
+	// column and is on the left
+
+	hashJoiner.hashTable = makeHashTableInt(hashTableBucketSize, len(hashJoiner.leftCols))
+
+	hashJoiner.build()
+}
+
+func (hashJoiner *hashJoinerIntInt) Next() dataFlow {
+	flow := hashJoiner.right.Next()
+	batchSize := flow.n
+
+	if batchSize == 0 {
+		return flow
+	}
+
+	batchProber := makeHashJoinProber(batchSize, hashJoiner.hashTable)
+	batchProber.start(flow, hashJoiner.rightEqColIdx, hashJoiner.rightCols, &hashJoiner.outFlow)
+
+	return hashJoiner.outFlow
+}
+
+// build performs the build phase of the hash join using the left relation.
+// Different builders used different heuristics for the build phase allowing us
+// to evaluate cpu-memory trade-offs.
```

Comment:
Great idea to have multiple builders! Perhaps we will be able to select a builder at plan time depending on the characteristics of the tables. Do you see any opportunities like that?

Reply:
Yes definitely :D Found some cool literature on how to optimize the build phase for various tradeoffs... and we will also need different builders/probers when we expand to N-N joins